pkg/server/plugin/notifier/gcpbundle/gcpbundle.go

```diff
@@ -0,0 +1,258 @@
+func (p *Plugin) NotifyAndAdvise(ctx context.Context, req *notifier.NotifyAndAdviseRequest) (*notifier.NotifyAndAdviseResponse, error) {
+	config, err := p.getConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	switch req.Event.(type) {
+	case *notifier.NotifyAndAdviseRequest_BundleLoaded:
+		// ignore the bundle presented in the request. see updateBundleObject for details on why.
+		if err := p.updateBundleObject(ctx, config); err != nil {
+			return nil, err
+		}
+	}
+	return &notifier.NotifyAndAdviseResponse{}, nil
+}
+
+func (p *Plugin) Configure(ctx context.Context, req *spi.ConfigureRequest) (resp *spi.ConfigureResponse, err error) {
+	if p.identityProvider == nil {
+		return nil, status.Error(codes.FailedPrecondition, "IdentityProvider host service is required but not brokered")
+	}
+
+	config := new(pluginConfig)
+	if err := hcl.Decode(&config, req.Configuration); err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "unable to decode configuration: %v", err)
+	}
+
+	if config.Bucket == "" {
+		return nil, status.Error(codes.InvalidArgument, "bucket must be set")
+	}
+	if config.ObjectPath == "" {
+		return nil, status.Error(codes.InvalidArgument, "object_path must be set")
+	}
+
+	p.setConfig(config)
+	return &spi.ConfigureResponse{}, nil
+}
+
+func (p *Plugin) GetPluginInfo(ctx context.Context, req *spi.GetPluginInfoRequest) (*spi.GetPluginInfoResponse, error) {
+	return &spi.GetPluginInfoResponse{}, nil
+}
+
+func (p *Plugin) getConfig() (*pluginConfig, error) {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+	if p.config == nil {
+		return nil, status.Error(codes.FailedPrecondition, "not configured")
+	}
+	return p.config, nil
+}
+
+func (p *Plugin) setConfig(config *pluginConfig) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	p.config = config
+}
+
+func (p *Plugin) updateBundleObject(ctx context.Context, c *pluginConfig) (err error) {
+	client, err := p.hooks.newBucketClient(ctx, c.ServiceAccountFile)
+	if err != nil {
+		return status.Errorf(codes.Unknown, "unable to instantiate bucket client: %v", err)
+	}
+	defer client.Close()
+
+	for {
+		// Get the config map so we can use the version to resolve conflicts racing
+		// on updates from other servers.
+		generation, err := client.GetObjectGeneration(ctx, c.Bucket, c.ObjectPath)
+		if err != nil {
+			return status.Errorf(codes.Unknown, "unable to get bundle object %s/%s generation: %v", c.Bucket, c.ObjectPath, err)
+		}
+		p.log.Debug("Bundle object generation retrieved", telemetry.Generation, generation)
+
+		// Load bundle data from the the identity provider. The bundle has to
+		// be loaded after fetching the config map so we can properly detect
+		// and correct a race updating the bundle (i.e. read-modify-write
+		// semantics).
+		resp, err := p.identityProvider.FetchX509Identity(ctx, &hostservices.FetchX509IdentityRequest{})
+		if err != nil {
+			st := status.Convert(err)
+			return status.Errorf(st.Code(), "unable to fetch bundle from identity provider: %v", st.Message())
```

Comment:
nit: this message feels like it could be a little friendlier if it referenced spire-server instead of identity provider because knowledge of the plugin implementation is needed in order to know what an/the identity provider is in this context

Reply:
👍